### PR TITLE
notehead (v5) extracted from note (v5)

### PR DESCRIPTION
**Pull request by rvilarl**
Minimal visual changes on stavenotes

### CODE REVIEW

**Comment by rvilarl** [comment]
current
Accidental Accidental_Arrangement_Special_Cases Bravura svg_current
reference
Accidental Accidental_Arrangement_Special_Cases Bravura svg_reference

**Comment by ronyeh** [comment]
Once these are merged, we should investigate how to streamline the "first time experience" for a new developer working with vexflow. Right now, if you don't have any of the music fonts installed, the grunt visual tests generate images like this:

image

Only the images with names starting with "canvas_" look correct. The others are either SVG or SVG-rendered-to-PNG, so they will be full of empty boxes unless the viewer has the SMuFL fonts installed.

File: src/tables.ts [hunk 1]
```diff
@@ -1049,254 +1093,214 @@ export class Tables {
...
-          default:
-            code = 'noteheadDiamondBlack';
-            break;
-        }
-        break;
-      case 'N':
-      case 'G':
+      /* Diamond */
+      case 'DI':
+      case 'H': // Harmonics
```
Comment (ronyeh) [inline src/tables.ts:1158]: Inconsistent comment style here. Should we just write `// Diamond`?

File: src/tables.ts [hunk 2]
```diff
@@ -729,6 +719,62 @@ export class Tables {
   static STAVE_LINE_THICKNESS = 1;
   static RENDER_PRECISION_PLACES = 3;
   static RESOLUTION = RESOLUTION;
+// 1/2, 1, 2, 4, 8, 16, 32, 64, 128
+  // NOTE: There is no 256 here! However, there are other mentions of 256 in this file.
+  // For example, in durations has a 256 key, and sanitizeDuration() can return 256.
+  // The sanitizeDuration() bit may need to be removed by 0xfe.
```
Comment (ronyeh) [inline src/tables.ts:725]: Needs an extra line before this line, and also some spaces before the `//`

File: src/tables.ts [hunk 3]
```diff
@@ -729,6 +719,62 @@ export class Tables {
...
+      codeFlagUp: '\ue246' /*flag64thUp*/,
+    },
+
+    128: {
+      beamCount: 5,
+      stemBeamExtension: 22.5,
+      stem: true,
+      codeFlagUp: '\ue248' /*flag128thUp*/,
+    },
+  };
```
Comment (ronyeh) [inline src/tables.ts:777]: Do we need to add a 256?

File: src/stavenote.ts [hunk 4]
```diff
@@ -532,10 +532,10 @@ export class StaveNote extends StemmableNote {
...
         stemDirection,
         customGlyphCode: noteProps.code,
-        glyphFontScale: this.renderOptions.glyphFontScale,
-        xShift: noteProps.shiftRight,
-        stemUpXOffset: noteProps.stemUpXOffset,
-        stemDownXOffset: noteProps.stemDownXOffset,
+        // #FIXME: HACK to use 30 as fontSize default rather than 39 used with glyphs.
+        // HACK-BEGIN
+        glyphFontScale: (this.renderOptions.glyphFontScale / 39) * 30,
+        // HACK-END
```
Comment (ronyeh) [inline src/stavenote.ts:538]: How will we resolve this hack later?

File: src/notehead.ts [hunk 5]
```diff
@@ -107,18 +43,20 @@ export class NoteHead extends Note {
...
-  protected stemUpXOffset: number = 0;
-  protected stemDownXOffset: number = 0;
   protected displaced: boolean;
   protected stemDirection: number;
 
   protected line: number;
   protected index?: number;
   protected slashed: boolean;
 
+  // map to correspoding glyph with ledger line
```
Comment (ronyeh) [inline src/notehead.ts:54]: Typo (correspoding) & formatting. 

Also... I'm not sure what this comment means. :-) :-(

**Review by ronyeh — CHANGES_REQUESTED** [review]
Looks OK. I just suggested some minor changes.

File: src/note.ts [hunk 6]
```diff
@@ -16,17 +16,21 @@ import { Category } from './typeguard';
...
+
 export interface KeyProps {
-  stemDownXOffset?: number;
-  stemUpXOffset?: number;
   key: string;
   octave: number;
   line: number;
   intValue?: number;
-  accidental?: string;
+  accidental?: number;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:32]: They are the SMuFL codes. It is an intermediate step, they will end up being strings in format '\u...'

File: src/note.ts [hunk 7]
```diff
@@ -92,6 +96,25 @@ export abstract class Note extends Tickable {
     return Category.Note;
   }
 
+  // Return a glyph given duration and type. The type can be a custom glyph code from customNoteHeads.
+  // The default type is a regular note ('n').
+  static getGlyphPropsNew(duration: string, type: string = 'n'): GlyphPropsNew {
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:101]: yes temporary for the extracted code.

File: src/note.ts [hunk 8]
```diff
@@ -217,6 +240,7 @@ export abstract class Note extends Tickable {
   // INSTANCE MEMBERS
 
   glyphProps: GlyphProps;
+  glyphPropsNew: GlyphPropsNew;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:243]: yes temporary

File: src/note.ts [hunk 9]
```diff
@@ -92,6 +96,25 @@ export abstract class Note extends Tickable {
...
+    // Lookup duration for default glyph head code
+    let code = Tables.durationCodes[duration];
+    if (code === undefined) {
+      code = Tables.durationCodes['4'];
+    }
+
+    // Try and get the note head
+    const codeNoteHead = Tables.codeNoteHead(type.toUpperCase(), duration);
+    // Merge duration props for 'duration' with the note head properties.
+    if (codeNoteHead !== '\u0000') code = { ...code, codeHead: codeNoteHead };
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:113]: done

File: src/note.ts [hunk 10]
```diff
@@ -424,7 +449,7 @@ export abstract class Note extends Tickable {
 
   /** Get the glyph width. */
   getGlyphWidth(): number {
-    return this.glyphProps.getWidth(this.renderOptions.glyphFontScale);
+    return 0;
```
[the remark this answers is not in the document]
Comment (rvilarl) [reply src/note.ts:452]: StaveNote:
```
/** Get the glyph width. */
getGlyphWidth(): number {
  return this.noteHeads[0].getWidth();
}
```

File: src/tables.ts (hunk 1 above)
Comment (rvilarl) [reply src/tables.ts:1158]: Done

File: src/tables.ts (hunk 2 above)
Comment (rvilarl) [reply src/tables.ts:725]: done

File: src/tables.ts (hunk 3 above)
Comment (rvilarl) [reply src/tables.ts:777]: Perhaps but this should be another PR. It is not a porting issue.

File: src/stavenote.ts (hunk 4 above)
Comment (rvilarl) [reply src/stavenote.ts:538]: The 39 / 30 will go away. This is needed for the extracted code.

File: src/notehead.ts (hunk 5 above)
Comment (rvilarl) [reply src/notehead.ts]: second try :)

**Review by ronyeh — APPROVED** [review]

**Pull request merged**